### PR TITLE
Client side checking

### PR DESCRIPTION
**Pull request by SteVio89**

### CODE REVIEW

File: languagetool-core/pom.xml [hunk 1]
```diff
@@ -347,6 +347,21 @@
             <artifactId>lombok</artifactId>
             <scope>provided</scope>
         </dependency>
+<!--        <dependency>-->
```
Comment (fabrichter) [inline languagetool-core/pom.xml:350]: is there a reason for this to stay in?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
+package org.languagetool.language.identifier;
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java:1]: Could you add the license header here?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.languagetool.language.identifier.detector.CommonWordsDetector;
+import org.languagetool.language.identifier.detector.UnicodeBasedDetector;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public abstract class LanguageIdentifier {
+    private static final Pattern URL_REGEX = Pattern.compile("https?://[-_.?&~;+=/#%0-9A-Za-z]+");   // '%' has been added
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java:18]: `protected` fields usually come before `private` ones, don't they?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    protected static final TextFilter REMOVE_EMAIL_SIGNATURE_FILTER = text -> SIGNATURE.matcher(text).replaceFirst("");
+    protected static final TextFilter REMOVE_MENTION_FILTER = text -> MENTION.matcher(text).replaceFirst("");
+    protected static final TextFilter REMOVE_NON_BREAKING_SPACES_FILTER = text -> text.toString().replace('\u00A0', ' ');
+    protected static final TextFilter REMOVE_URL_FILTER = text -> MAIL_REGEX.matcher(URL_REGEX.matcher(text).replaceAll(" ")).replaceAll(" ");
+
+    protected static final UnicodeBasedDetector UNICODE_BASED_LANG_IDENTIFIER = new UnicodeBasedDetector();
+    protected static final CommonWordsDetector COMMON_WORDS_LANG_IDENTIFIER;
+
+    static {
+        try {
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java:35]: Indent should be 2 spaces, as usual

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
+package org.languagetool.language.identifier;
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java:1]: + license header

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import org.jetbrains.annotations.TestOnly;
+import org.languagetool.Languages;
+
+import java.io.File;
+import java.util.List;
+
+@Slf4j
+public enum LanguageIdentifierService {
+
+    INSTANCE;
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java:14]: 2-space indent

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+     * @return new {@code LanguageIdentifier} or existing if already initialized.
+     * @since 5.8
+     */
+    public synchronized LanguageIdentifier getSimpleLanguageIdentifier(@Nullable List<String> preferredLangCodes) {
+        if (simpleIdenfifier == null) {
+            if (preferredLangCodes == null) {
+                this.simpleIdenfifier = new SimpleLanguageIdentifier();
+            } else {
+                this.simpleIdenfifier = new SimpleLanguageIdentifier(preferredLangCodes);
+
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java:54]: whitespace line

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            defaultIdentifier.enableNgrams(ngramLangIdentData);
+            defaultIdentifier.enableFasttext(fasttextBinary, fasttextModel);
+            this.defaultIdentifier = defaultIdentifier;
+        }
+        return this.defaultIdentifier;
+    }
+
+    /**
+     * @param preferredLangCodes - a list of language codes for that a spellchecker will be initialized.
+     *                           If {@code null} all spellchecker will be used.
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifierService.java:44]: all spellchecker**s**

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
+package org.languagetool.language.identifier;
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:1]: + license header

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                spellingRuleLanguage = language.getDefaultLanguageVariant();
+            }
+            if (spellingRuleLanguage == null) {
+                spellingRuleLanguage = language;
+            }
+            SpellingCheckRule defaultSpellingCheckRule = spellingRuleLanguage.getDefaultSpellingRule(JLanguageTool.getDataBroker().getResourceBundle(JLanguageTool.MESSAGE_BUNDLE, new Locale(spellingRuleLanguage.getShortCodeWithCountryAndVariant())));
+            if (defaultSpellingCheckRule != null) {
+                spellingCheckRules.put(language.getShortCode(), defaultSpellingCheckRule);
+            } else {
+                log.warn("Could not found default speller rule for {}", language.getShortCode());
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:40]: "Could not find"

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        }
+    }
+
+    public SimpleLanguageIdentifier(List<String> preferredLangCodes) {
+        super(1000);
+        log.info("Init SpellcheckLangIdentifier with {}", preferredLangCodes);
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:47]: SpellcheckLangIdentifier -> SimpleLanguageIdentifier ?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        List<String> additionalLangs = langPair.getLeft();
+        List<String> preferredLangs = langPair.getRight();
+
+        String[] words = cleanText.split(" ");
+        List<String> dominantLangCodes = UNICODE_BASED_LANG_IDENTIFIER.getDominantLangCodes(cleanText);
+        Map<String, Double> scores = new HashMap<>();
+        String detectionSource = "spellchecker";
+        for (Map.Entry<String, SpellingCheckRule> scp : spellingCheckRules.entrySet()) {
+            if (dominantLangCodes.contains(scp.getKey()) ^ (dominantLangCodes.isEmpty() && !NON_LATIN_CHARS_LANGUAGES.contains(scp.getKey()))) {
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:72]: `^` ?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                double errors = 0;
+                for (String word : words) {
+                    try {
+                        errors += scp.getValue().isMisspelled(word) ? 1 : 0;
+                    } catch (IOException ex) {
+                        log.error("Got {}", ex.getMessage());
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:78]: Should we really log this instead of just throwing the exception?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+                    }
+                }
+                double errorRate = errors / (double) words.length;
+                log.info("Found {} errors for {} words with {} spellchecker this scores in an error rate of {}", errors, words.length, scp.getKey(), errorRate);
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:82]: comma missing between "spellchecker this"

File: languagetool-core/src/main/java/org/languagetool/language/identifier/detector/UnicodeBasedDetector.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+public class UnicodeBasedDetector {
+
+  private static final int DEFAULT_MAX_CHECK_LENGTH = 50;
+  private static final float THRESHOLD = 0.5f;
+
+    private final int maxCheckLength;
+
+  public UnicodeBasedDetector() {
+    this(DEFAULT_MAX_CHECK_LENGTH);
+    }
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/detector/UnicodeBasedDetector.java:33]: incoherent indentation

File: languagetool-server/src/main/java/org/languagetool/server/HTTPServerConfig.java [hunk 16]
```diff
@@ -456,6 +461,12 @@ private void parseConfigFile(File file, boolean loadLangModel, boolean loadWord2
         }
         slowRuleLoggingThreshold = Integer.valueOf(getOptionalProperty(props, "slowRuleLoggingThreshold", "-1"));
         disabledRuleIds = Arrays.asList(getOptionalProperty(props, "disabledRuleIds", "").split(",\\s*"));
+        localApiMode = Boolean.parseBoolean(getOptionalProperty(props, "localApiMode", "false"));
+        motherTongue = getOptionalProperty(props, "motherTongue", "en-US");
+        String preferredLanguages = getOptionalProperty(props, "preferredLanguages", "").replace(" ", "");
+        if (preferredLanguages != "") {
+            this.preferredLanguages = Arrays.asList(preferredLanguages.split(","));
```
Comment (danielnaber) [inline languagetool-server/src/main/java/org/languagetool/server/HTTPServerConfig.java:468]: 2-space indent

File: languagetool-standalone/src/test/java/org/languagetool/language/LangaugeIdentifierServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
+package org.languagetool.language;
```
Comment (danielnaber) [inline languagetool-standalone/src/test/java/org/languagetool/language/LangaugeIdentifierServiceTest.java:1]: +license header

File: pom.xml [hunk 18]
```diff
@@ -226,8 +226,8 @@
                     </executions>
                 </plugin>
                 <plugin>
-                    <groupId>pl.project13.maven</groupId>
-                    <artifactId>git-commit-id-plugin</artifactId>
+                    <groupId>io.github.git-commit-id</groupId>
```
Comment (danielnaber) [inline pom.xml:229]: Is this actually related to this change?

**Review by danielnaber — CHANGES_REQUESTED** [review]

File: languagetool-core/pom.xml (hunk 1 above)
Comment (SteVio89) [reply languagetool-core/pom.xml:350]: No, I remove it with the next commit.

File: pom.xml (hunk 18 above)
Comment (SteVio89) [reply pom.xml:229]: Yes, we need this Version if we want to use JDK17, which is needed for the jpackage.

File: languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java (hunk 3 above)
Comment (SteVio89) [reply languagetool-core/src/main/java/org/languagetool/language/identifier/LanguageIdentifier.java:18]: We got an `illegal forward reference` if we move the protected fields before the private ones.

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+                scores.put(scp.getKey(), 1.0 - errorRate);
+            }
+        }
+        if (scores.isEmpty()) {
+            scores.put("zz", 1.0);
+        }
+        log.info("Got scores: {}", scores);
```
Comment (fabrichter) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:89]: can this be log.debug?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java (hunk 12 above)
Comment (SteVio89) [reply languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:72]: Javas XOR operator → Only check the language with this spellchecker if:
a) the spellchecker is from a language with non-Latin chars, AND the text contain non-Latin chars → **is true**
OR
b) The spellchecker is not from a language with non-Latin chars, AND the text does not contain any non-Latin chars.→ **is true**

Does not match if both are true, as a normal `or` would do. Could also be replaced by `!=`

Known issue:
Leads to problems with texts that contain non-Latin chars, but actually have a latin language as the main language. These would be misidentified.

**Review by fabrichter — APPROVED** [review]

File: languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+    Map<String, Double> scores = new HashMap<>();
+    String detectionSource = "spellchecker";
+    for (Map.Entry<String, SpellingCheckRule> scp : spellingCheckRules.entrySet()) {
+      if (dominantLangCodes.contains(scp.getKey()) ^ (dominantLangCodes.isEmpty() && !NON_LATIN_CHARS_LANGUAGES.contains(scp.getKey()))) {
+        double errors = 0;
+        for (String word : words) {
+          try {
+            errors += scp.getValue().isMisspelled(word) ? 1 : 0;
+          } catch (IOException ex) {
+            throw new RuntimeException(ex.getMessage());
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/SimpleLanguageIdentifier.java:97]: Should be `ex` instead of `ex.getMessaget()` so we don't lose information?

File: languagetool-core/src/main/java/org/languagetool/language/identifier/detector/UnicodeBasedDetector.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class UnicodeBasedDetector {
+
+  private static final int DEFAULT_MAX_CHECK_LENGTH = 50;
+  private static final float THRESHOLD = 0.5f;
+
+    private final int maxCheckLength;
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/language/identifier/detector/UnicodeBasedDetector.java:29]: wrong indent

File: languagetool-language-modules/sv/src/main/java/org/languagetool/language/Swedish.java [hunk 22]
```diff
@@ -83,4 +85,10 @@ public List<Rule> getRelevantRules(ResourceBundle messages, UserConfig userConfi
     );
   }
 
+  @Nullable
+  @Override
+  protected SpellingCheckRule createDefaultSpellingRule(ResourceBundle messages) throws IOException {
+    return new HunspellRule(messages, this, null, null);
+
```
Comment (danielnaber) [inline languagetool-language-modules/sv/src/main/java/org/languagetool/language/Swedish.java:92]: empty line

File: languagetool-server/src/main/java/org/languagetool/server/HTTPServerConfig.java [hunk 23]
```diff
@@ -1466,5 +1477,19 @@ public void setRedisKeyPassword(String redisKeyPassword) {
...
+      return localApiMode;
+  }
+
+  public String getMotherTongue() {
+      return motherTongue;
+  }
+
+  public List<String> getPreferedLanguages() {
+      return preferredLanguages;
+  }
```
Comment (danielnaber) [inline languagetool-server/src/main/java/org/languagetool/server/HTTPServerConfig.java:1494]: 2-space indent

File: languagetool-standalone/src/test/java/org/languagetool/language/SimpleLanguageIdentifierTest.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+import org.languagetool.language.identifier.LanguageIdentifierService;
+
+public class SimpleLanguageIdentifierTest extends LanguageIdentifierTest {
+
+    private LanguageIdentifier identifier = LanguageIdentifierService.INSTANCE
+                .clearLanguageIdentifier("simple")
+                .getSimpleLanguageIdentifier(null);
+    
+    @Test
+    public void testDetection() {
```
Comment (danielnaber) [inline languagetool-standalone/src/test/java/org/languagetool/language/SimpleLanguageIdentifierTest.java:15]: indent
